protocol/implementations/js/src/types.ts

```diff
@@ -1,32 +1,76 @@
-export namespace TbDex {
-  export type Offering = {
-    placeholder: string;
-  }
-  
-  export type RFQ = {
-    test: string;
-  }
-  
-  export type Quote = {
-    placeholder: string;
-  }
-  
-  export type MessageTypes = {
-    rfq: RFQ,
-    quote: Quote
-  }
-  
-  export type MessageType<M extends keyof MessageTypes> = MessageTypes[M];
-  
-  export type Message<T extends keyof MessageTypes> = {
-    id: string;
-    contextId: string;
-    from: string;
-    to: string;
-    type: T;
-    body: MessageTypes[T];
-    dateCreated: string;
+export type TbDEXMessage = Metadata &
+  (
+    | {
+        type: "offering"
+        body: Offering
+      }
+    | {
+        type: "rfq"
+        body: Rfq
+      }
+    | {
+        type: "quote",
+        body: Quote
+      }
+    | {
+        type: "order"
+        body: Order
+      }
+    | {
+        type: "orderStatus"
+        body: OrderStatus
+      }
+  )
+
+export interface Metadata {
+  id: string
+  contextId: string
+  from: string
+  to: string
+  createdTime: string
+}
+export interface Offering {
+  description: string
+  pair: string
+  unitPrice: string
+  baseFee?: string
+  min: string
+  max: string
+  presentationRequestJwt: string
+  payinInstruments: PaymentInstrument[]
+  payoutInstruments: PaymentInstrument[]
+}
+export interface PaymentInstrument {
+  kind: "DEBIT_CARD" | "BITCOIN_ADDRESS"
+  fee?: {
+    flatFee?: string
   }
 }
-
-
+export interface Rfq {
+  pair: string
+  amount: string
+  verifiablePresentationJwt: string
+  payinInstrument: PaymentInstrument
+  payoutInstrument: PaymentInstrument
+}
+export interface Quote {
+  expiryTime: string
+  totalFee: string
+  amount: string
+  paymentPresentationRequestJwt: string
+  paymentInstructions: PaymentInstructions
+}
+export interface PaymentInstructions {
+  payin?: PaymentInstruction
+  payout?: PaymentInstruction
+}
+export interface PaymentInstruction {
+  link?: string
+  instruction?: string
+}
+export interface Order {
+  paymentVerifiablePresentationJwt: string
+}
+export interface OrderStatus {
+  orderStatus: "PENDING" | "COMPLETED" | "FAILED"
```

Comment:
Is there any difference between this and using an `enum` type in TS?

Reply:
mm maybe not but i like enums so i added that back in. i wonder if there was a way to say "make this output enums" in the schema...

Reply:
Ya i'm surprised the `enum` json schema type doesn't do that for you? >_<

Reply:
my intuition is just build objects for enums, like this...

```typescript
export const Types = {
  Offering    : 'Offering',
  RFQ         : 'RFQ',
  Quote       : 'Quote',
  Order       : 'Order',
  OrderStatus : 'OrderStatus'
};
```

ts enums are terrible